src/metrics.jl

```diff
@@ -236,7 +236,7 @@ Base.@propagate_inbounds function _evaluate(d::UnionMetrics, a::AbstractArray, b
     end
     @inbounds begin
         s = eval_start(d, a, b)
-        if IndexStyle(a, b) === IndexLinear() || size(a) == size(b)
+        if IndexStyle(a, b) === IndexLinear() && axes(a) == axes(b)
```

Comment:
I believe it should still be `||` here.

An example of a non-linear indexing array is `PaddedView`:

```julia
x = rand(2, 3);
px = PaddedView(-1, x, (4, 4));
IndexStyle(px); # IndexCartesian()
axes(px); # (Base.OneTo(4), Base.OneTo(4))

y = rand(3, 2);
py = PaddedView(-1, y, (4, 4));
axes(py) # (Base.OneTo(4), Base.OneTo(4))

axes(py) == axes(px) # true
```

In this case, they could share the same index `I` in the loop and hence we don't need to use the fallback implementation, where axes of arrays are different.

Reply:
If I put `||`, then the `OffsetArray` fix won't work, because the example in #93 has linearly indexed vectors. I'll go with
```julia
if (IndexStyle(a, b) === IndexLinear() && eachindex(a) == eachindex(b)) || axes(a) == axes(b)
```
This would catch the different indices with OffsetArrays, but allow cartesian-indexing as long as the axes are the same.

Reply:
With this PR, you can do
```julia
Euclidean()(px, py)
```
and it works. According to the latest logic, it uses the `@simd` branch.